procurement_mto_analytic/models/stock_rule.py

```diff
@@ -6,22 +7,25 @@
 class StockRule(models.Model):
     _inherit = "stock.rule"
 
-    def _make_po_get_domain(self, company_id, values, partner):
-        domain = super()._make_po_get_domain(company_id, values, partner)
-        if not values.get("analytic_distribution"):
-            return domain
-        matching_lines = (
-            self.env["purchase.order"]
-            .sudo()
-            .search([dom for dom in domain])
-            .order_line.filtered(
-                lambda x: x.analytic_distribution == values.get("analytic_distribution")
-            )
-        )
-        if matching_lines:
-            domain += (("order_line", "in", tuple(matching_lines.ids)),)
-        else:
-            domain += (
-                ("order_line", "=", False),
-            )  # This might need adjustment based on your intent.
-        return domain
+    def _get_po_line_grouping_key(self, line):
+        """
+        Extend the grouping logic to conditionally include the analytic_distribution
+        only if it exists.
+        """
+        original_key = super()._get_po_line_grouping_key(line)
+        if line.analytic_distribution:
+            analytic_distribution_key = tuple(line.analytic_distribution.items())
```

Comment:
> if I have a distribution with the same accounts and the same percentages, those should be the same regardless of the order, right? I suggest adding a sorted call into the key generation

@aisopuro items of `analytic_distribution` are always sorted by default. So, we don't need to worry about it.

Reply:
Could you link to the place in Odoo's source code that ensures they are always sorted, regardless of whether:
* The distribution was added by a user through the UI
* The distribution was added by business logic in Python
* The distribution was imported via excel
* etc

Or does the `JSONB` column type in Postgresql guarantee keys are ordered?

Reply:
Depends on this source(https://www.postgresql.org/docs/10/datatype-json.html#JSON-TYPE-MAPPING-TABLE:~:text=There%20are%20two,value%20is%20kept.), I conclude analytic_distribution are sorted depends on the efficiency.
However we insert the key in any order, it will be sorted if they have same data.
Eg, when we insert {"1":60, "2":40} and {"2":40, "1":60}, they will be sorted. So, the result will be the same.
<img width="1117" alt="Screenshot 2024-05-17 at 10 23 02" src="https://github.com/OCA/account-analytic/assets/45355704/dc458f44-d5f2-4dce-b772-41271db0484e">

Reply:
Nice :+1: